Fix Eigen5 support on Pinocchio 3

jcarpent opened the pull request:
Thanks for taking the time to make and fill out this pull request!

## Description

<!--
Please include a clear and concise description of the changes included in this pull request.
Explain what are you changing and why.

If applicable, link the related issue using "Fixes #issue_number".
-->

## Checklist

- [x] I have run `pre-commit run --all-files` or `pixi run lint`
- [x] I have performed a self-review of my own code
- [ ] I have commented my code where necessary
- [ ] I have made corresponding changes to the doxygen documentation
- [x] I have added tests that prove my fix or feature works
- [x] I have updated the [CHANGELOG](https://github.com/stack-of-tasks/pinocchio/blob/devel/CHANGELOG.md) or added the "no changelog" label if it's a CI-related issue
- [ ] I have updated the [README credits section](https://github.com/stack-of-tasks/pinocchio?tab=readme-ov-file#credits)

jcarpent committed:
core: fix return type issue with Eigen >= 3.4.90

jcarpent committed:
core: remove deprecated EIGEN_NOEXCEPT

jcarpent committed:
changelog: sync

jcarpent commented:
@nim65s Following a recommendation made by @jorisv, could you please add the support of Eigen5 in Nix?

nim65s committed:
Nix: add job with eigen5

nim65s commented:
I've added a nix job with eigen 5, without python, casadi & coal.
I've also checked that it is failing as expected if a `EIGEN_NOEXCEPT` is not replaced by `noexcept`:
```cpp
In file included from /build/source/unittest/tridiagonal-matrix.cpp:7:
/build/source/include/pinocchio/math/tridiagonal-matrix.hpp:263:41: error: expected ';' at end of member declaration
  263 |     EIGEN_CONSTEXPR Eigen::Index rows() const EIGEN_NOEXCEPT
      |                                         ^~~~~
      |                                              ;
/build/source/include/pinocchio/math/tridiagonal-matrix.hpp:263:47: error: 'EIGEN_NOEXCEPT' does not name a type
  263 |     EIGEN_CONSTEXPR Eigen::Index rows() const EIGEN_NOEXCEPT
      |                                               ^~~~~~~~~~~~~~
```

jorisv committed:
tests: Fix spatial test with Eigen5

Random number generation algorithm has changed in Eigen5.
Some non robust tests are not passing because of that.

jorisv reviewed (changes requested):
Casadi support is not working

jorisv committed:
casadi: Fix build with Eigen5

jorisv commented:
Casadi support is fixed now.
I can't test easily python binding since I would have to rebuild eigenpy.
Do we merge even if the support is partial ?

nim65s commented:
as discussed in internal meeting, this will wait for pinocchio v4
